feat(k8s): add Helm Chart

apricote opened the pull request:
##  What this PR does

- Adds a helm chart for the hcloud-csi driver (Closes #108 #369)
- Remove the old kustomization based manifests in favor of the Helm Chart
- Use the Helm Chart to render the static manifests that are currently used for install
- Publish the Helm Chart to https://github.com/hetznercloud/helm-charts

## Reviewer Notes

This replaces #398 as I did not have push permissions for the branch.

I recommend reviewing by commit, to reduce the noise. This is still a huge PR and if requested I can split it up into multiple.

A lot of the scripts and goreleaser config was shamelessly stolen from hcloud-cloud-controller-manager.

jooola commented on chart/templates/_common_images.tpl:
```diff
@@ -0,0 +1,79 @@
+{{/*
+Return the proper image name
+{{ include "common.images.image" ( dict "imageRoot" .Values.path.to.the.image "global" .Values.global "all" $ ) }}
+*/}}
+{{- define "common.images.image" -}}
+{{- $registryName := .imageRoot.registry -}}
+{{- $repositoryName := .imageRoot.repository -}}
+{{- $separator := ":" -}}
+{{- $termination := tpl (.imageRoot.tag | toString) .all -}}
+{{- if .global }}
+    {{- if .global.imageRegistry }}
+     {{- $registryName = .global.imageRegistry -}}
+    {{- end -}}
+{{- end -}}
+{{- if .imageRoot.digest }}
+    {{- $separator = "@" -}}
+    {{- $termination = .imageRoot.digest | toString -}}
+{{- end -}}
+{{- if $registryName }}
+    {{- printf "%s/%s%s%s" $registryName $repositoryName $separator $termination -}}
+{{- else -}}
+    {{- printf "%s%s%s"  $repositoryName $separator $termination -}}
+{{- end -}}
+{{- end -}}
```
Is this really required ? This feels really overkill, while the user could define their full image URL themselves in a single `image` field, with tag and/or digest, registry if they need it.

apricote replied:
I switched to a single string field.

jooola commented on chart/templates/_common_images.tpl:
```diff
@@ -0,0 +1,79 @@
+{{- end -}}
+
+{{/*
+Return the proper Docker Image Registry Secret Names (deprecated: use common.images.renderPullSecrets instead)
+{{ include "common.images.pullSecrets" ( dict "images" (list .Values.path.to.the.image1, .Values.path.to.the.image2) "global" .Values.global) }}
```
If this is deprecated, maybe we can remove the function instead ? Or replace the new logic and keep the name to keep it aligned with the rest.

I believe the example is invalid as it contains `,` to separate the list items.

apricote replied:
Makes sense, did not even see this. Removed this one and shifted all usage to the `renderPullSecrets` template (while renaming it to `pullSecrets`

jooola reviewed (changes requested):
I did not finish my review, but here is already a few points:

- Could we add tests for this ? The templating mixed with yaml feels a bit messy and I fear that it might hide a bunch of bugs. What about generating a set of charts with different `.Values` and compare it to some snapshots we keep in the repo?
- Could we reduce the complexity by writing good documentation instead of making the chart overly complex and handling every last field (e.g. not build the image field) ?
- Some files have an inconsistent naming, for example pbd.yaml or extra-list.yaml. I would also be in favor of using kebab case filenames.

apricote commented on chart/README.md:
```diff
@@ -0,0 +1,59 @@
+
+This Helm chart is the recommended installation method for [hcloud-csi-driver](https://github.com/hetznercloud/csi-driver) on Kubernetes.
+
+> [!WARNING]
+> The Helm Chart is not yet published and the instructions below will not work until the next release (v2.5.0).
```
I did not want to force our hands one this. I was planning on deploying the Chart to my personal cluster before creating a release, but I can also do that from this PR.

apricote commented:
> Could we add tests for this ? The templating mixed with yaml feels a bit messy and I fear that it might hide a bunch of bugs. What about generating a set of charts with different .Values and compare it to some snapshots we keep in the repo?

For sure, we already do that with the existing deployment manifest. I added two additional snapshots: default values & production example.

> Some files have an inconsistent naming, for example pbd.yaml or extra-list.yaml. I would also be in favor of using kebab case filenames.

Removed `extra-list.yaml` and the associated functionality (deploying random resources?). Renamed `pdb.yaml`. Not so sure about the kebap-case, never seen that in a helm chart :D

> Should we introduce a values.schema.json file to validate the values files ?

Sounds good :) In the latest push.

apricote commented on chart/.snapshots/example-prod.yaml:
```diff
@@ -0,0 +1,437 @@
+            - name: HCLOUD_VOLUME_DEFAULT_LOCATION
+              value: "nbg1"
+            - name: ENABLE_METRICS
+              value: "false"
```
Enabled :heavy_check_mark:

apricote commented on chart/templates/_common_images.tpl:
```diff
@@ -0,0 +1,27 @@
+{{/*
+Return the proper Container Image Registry Secret Names evaluating values as templates
+{{ include "common.images.pullSecrets" ( dict "images" (list .Values.path.to.the.image1, .Values.path.to.the.image2) "context" $) }}
```
Updated & also found a bug in how this template is called that caused the image specific pullSecrets to not show up in the template.

jooola replied:
Hmm, I am not sure if we cover these templates with the snapshots tests. Feel free to add a pull secret if they are missing.

apricote commented on chart/values.yaml:
```diff
@@ -0,0 +1,692 @@
+  ## @param controller.pdb.minAvailable Set minAvailable for controller pods
+  ## @param controller.pdb.maxUnavailable Set maxUnavailable for controller pods
+  ##
+  pdb:
```
Done :heavy_check_mark:

jooola commented:
Nice!

apricote commented:
I added an additional snapshot test that defines all optional values to make sure everything works. While doing that I found a bunch of features that did not work properly and fixed those.

apricote commented:
Manual testing went well. I also updated my personal cluster to use the chart from this branch and everything worked as expected :)

jooola commented on chart/README.md:
````diff
@@ -29,7 +29,7 @@ If you're unfamiliar with Helm it would behoove you to peep around the documenta
 
 ```sh
-kubectl apply -f https://raw.githubusercontent.com/hetznercloud/csi-driver/latest/deploy/kubernetes/hcloud-csi.yml
+kubectl apply -f https://raw.githubusercontent.com/hetznercloud/csi-driver/v2.4.0/deploy/kubernetes/hcloud-csi.yml
````
Shouldn't the version also be applied in the command below ?

apricote replied:
Fixed.

jooola reviewed (approved):
Great effort to build the snapshots tests! :100:
